### ISSUE TITLE
MacOS reseting zoom on start

### ISSUE THREAD

**Issue by Ctorum** [pragtical/pragtical#34]
I have a MacBook Air M1, and every time I launch the app, it starts zoomed in. I need to zoom out every time I start the app again. It look like this:

<img width="1440" alt="image" src="https://github.com/pragtical/pragtical/assets/66646826/85ce0841-0685-43c4-879b-b9244dfd9e00">

After the zoom out:

<img width="1440" alt="image" src="https://github.com/pragtical/pragtical/assets/66646826/7ec28798-542e-4b2f-bdf7-186791f6913c">

Is there any configuration I need to do? I don't want to zoom out every single time I reopen the app.

**Comment by jgmdev** [comment]
Settings -> Plugins -> Scale
![pragtical-scale](https://github.com/pragtical/pragtical/assets/1702572/feabe1c6-dd14-45f4-843f-aade7bad1e78)

By the way how is your M1 display configured? What resolution, scale factor?

**Comment by Ctorum** [comment]
I like the native settings and I like to use my Mac without other displays, that's why I barely changed any display config. Even so, my resolution is 2560 x 1600 with a 16:10 ratio.

**Comment by vcambur** [comment]
Had the same issue which was fixed by setting scale to 100%. However status bar's first message still has a wrong scale. Subsequent messages are ok. My resolution is set to 1680x1050 (16:10). No external monitors attached.

@jgmdev thank you for your work and time
<img width="1393" alt="Screenshot 2023-11-27 at 09 35 19" src="https://github.com/pragtical/pragtical/assets/10867337/30778ec3-3269-4040-bdf1-8e17ec463fc7">

**Comment by jgmdev** [comment]
> However status bar's first message still has a wrong scale.

Fixed that one on #103

**Comment by LeDominik** [comment]
This one is tricky tbh... so I have a MBP M3 Pro 16" and _this_ is what auto-detect looks like:

![image](https://github.com/user-attachments/assets/4e28ee41-c552-4c43-b7c5-f4a194f26f1d)

So fonts and other distances somehow don't fit together

Now this is with the default setting of MacOS, so it's traditional 2x retina from native resolution of 3456 × 2234
![image](https://github.com/user-attachments/assets/87ddbf95-29fa-478e-ab78-704560367913)

When I go to 200% zoom which should be "correct", it remains ugly:
![image](https://github.com/user-attachments/assets/d0af4fba-812a-4d8d-b5d6-84f9dacc2f2d)

When I attach an external monitor at work which is non-retina / HiDPI and move the window there it looks just as it should at 200%:
![image](https://github.com/user-attachments/assets/f9cffebd-769b-42e6-ba05-d322259fe725)

Although it's at native resolution and should thus be more correct at 100% but hey...
![image](https://github.com/user-attachments/assets/9570d98a-2587-4bb4-8b45-5b0a0a9966c9)

Long story short: works great on the external screen but somehow not on the local. I will retry after a reboot without any monitors attached maybe it's different at first...

**Comment by jgmdev** [comment]
Can you test if there is a different behavior building from source with the SDL hardware renderer backend disabled by editing:

**meson.build**
```diff
- if get_option('renderer') or host_machine.system() == 'darwin'
+ if get_option('renderer')
```

then build with something like:
```
meson setup --wrap-mode=forcefallback -Dppm=false -Drenderer=false build
meson compile -C build
```

and finally run it to test with:
```
./scripts/run-local -global build
```

This could help determine if a recently added change that performs internal font scaling with `renderer` is causing the issue.

**Comment by LeDominik** [comment]
Ok, somewhere I'm too stupid :-)

Build went through just a few warnings but I don't quite get how it's supposed to start (also tried to understand the script)

```
[650/652] Linking target src/pragtical
ld: warning: ignoring duplicate libraries: '-liconv'
[652/652] Linking static target subprojects/pcre2-10.44/libpcre2-32.a

~/tmp/pragtical-3.4.2 is 📦 v3.4.2 took 26s 
❯ ./scripts/run-local -global build
realpath: .run: No such file or directory
```

**Comment by jgmdev** [comment]
> ```
> ...
> realpath: .run: No such file or directory
> ```
> there's no `.run` dir anywhere... I do have the binary on `build/src/pragtical` but that then doesn't find it's `start.lua`

That one can be fixed by installing the `realpath` command provided by brew, If I recall correctly it was `brew install coreutils`

> BUT it's no longer usable, the mouse is not detected where it actually is -- see the highlight in the following screenshot in which the mouse is actually on the configuration icon way below...but it **thinks** it's on the meson.build file

You actually discovered some regressions on scaling code which I didn't notice after merging newer font loading/caching mechanism. This should serve me as a reminder of how important is to test different scalings after adding code that touches the rendering engine.

Could you give #134 a try and report how it behaves with those changes?

You can download https://patch-diff.githubusercontent.com/raw/pragtical/pragtical/pull/134.diff to your local pragtical git directory and run:

```sh
git apply 134.diff
```

Then recompile and run, just don't forget to `git restore meson.build` or turn the renderer on with `meson configure -Drenderer=true build` before `meson compile -C build`

I tested on my linux machine with wayland and fractional scaling and it seems better, sadly my Mac Book Air doesn't has a retina display, so not sure how it will behave with that :(

**Comment by LeDominik** [comment]
Hej; now it's working; I'd still say that font size vs. space around it is not perfect and it starts too big. 

Immediately after starting with only internal (retina) screen it starts huge, so autodetect seems to be thinking 200%:

![image](https://github.com/user-attachments/assets/e6b46ab4-d531-48d7-947d-4330909c4664)

After setting to 100% it becomes very useable (as can be seen in proportion to title bar etc...)

![image](https://github.com/user-attachments/assets/c1490ed3-d212-4405-8143-8fc381c9ee8d)

Regarding realpath - I did have it installed, so really it didn't find stuff; I'm now always installing into a prefix defined beforehand and then running from there with `bin/pragtical` but that works.

**Comment by jgmdev** [comment]
Ok, made another set of changes on #134:

* Disabled the "hardware renderer" by default for macOS because the fonts on scales >=2 seems to be off (which looks ugly) with latest glyph loading optimizations (this needs to be properly inspected later to find where the calculations are off, tested lite-xl master at 2x scale and it exhibits the same issue)
* Improved scale autodetection handling on the default software renderer, also improved it for the "hardware renderer" (it will now respect the internal scaling so it shouldn't start with a huge font size :smile:).
* Fixed emptyview icon initialization not in place when starting with non default scale
* Various improvements to scale plugin for better behavior and respecting of PRAGTICAL_SCALE environment variable. Also changed the settings a bit but, now I have to fix the numberbox widget to release clicks on scale changes :roll_eyes:

![scale-settings](https://github.com/user-attachments/assets/9e7e91a3-11e4-4149-a033-08cbbfe4c97c)

I tested locally on Linux/Wayland with fractional scaling and integral 2x scaling but now we need the macOS retina test :pleading_face:

**Comment by LeDominik** [comment]
Ok, I'll do my test on master :-)

**Comment by LeDominik** [comment]
Hej just built on `master` completely fresh and into an empty dir, so it's autoconfig all the way through.

```
meson setup --buildtype=release --wrap-mode=forcefallback -Dppm=false --prefix ~/tmp/prag build
meson compile -C build
meson install -C build
cd ~/tmp/prag
bin/pragtical
```

And it's just spot-on:

![image](https://github.com/user-attachments/assets/fb1bd5ae-eaae-4dc9-aee1-53738d18457a)

And here's the hardcore test

- I started pragtical on the retina laptop screen -> ok (see Screenshot)
- Moved it to the exteral _non retina_ -> ok, also stable on resizing
- Moved back to retina -> got supersmall

But that's really nitpicky, I don't know if you even have control around that. But font rendering and everything is just nicely centered, etc...

**Comment by jgmdev** [comment]
> And it's just spot-on:

Wohoo! :partying_face: that means we can now run on software renderer on macOS which is faster than the half baked hacky "hardware renderer" which was not really hardware rendering since it is still doing software rendering and performing an extra copy of final re-scaled output surface (which needed recalculating all rects) into a texture. With the mechanism we are using now the editor is performing the scale once for us each time the scale changes on the displaychanged event, the only drawback would be if you have the editor open and change current display scale to a different one, the editor will not detect that :cry:

> * Moved back to retina -> got supersmall

Hope my assumptions are right and #143 fixes that one. If it does we should be on a pretty good state for a bug fix release :D, when you have the time please test it and let me know.

**Comment by LeDominik** [comment]
Sorry for not much feedback - I can do a Retina + 4K test anyday, Office currently closed for some construction work so I can really only retest in the funky MBP Retina + normal resolution external around Tuesday...

**Comment by jgmdev** [comment]
Understandable, I hope v3.4.4 should be now on a better state. I tested on Gnome Wayland/X11 at 2X scale (switching back and forward between 1X - 2X while pragtical was open), also KDE and  Wayfire + XFCE with fractional scaling. Just tested windows and discovered that needed these changes #158 to prevent OS autoscaling (which brings blurriness). Haven't tested on 4k display or second display with different scale, have the equipment to test but been lazy to move stuff around...

Hopefuly the retina test from your side goes well!

**Comment by jgmdev** [comment]
Closing this as scale logic should be on better shape now. Thanks to everyone involved.
